### PR TITLE
Osi/balancer

### PR DESCRIPTION
**Pull request by 0xOsiris**
Initial Balancer V2 Integration
- Balancer Factory Structure
- Batch Request Module
- `BalancerV2Pool` Structure

## PR Checklist

- [ ] Added Tests
- [ ] Added Documentation
- [ ] Breaking changes

### CODE REVIEW

File: src/amm/balancer_v2/mod.rs [hunk 1]
```diff
@@ -0,0 +1,175 @@
...
+    #[derive(Debug, PartialEq, Eq)]
+    #[sol(rpc)]
+    contract IUniswapV2Pair {
+        event Sync(uint112 reserve0, uint112 reserve1);
+        function getReserves() external view returns (uint112 reserve0, uint112 reserve1, uint32 blockTimestampLast);
+        function token0() external view returns (address);
+        function token1() external view returns (address);
+        function swap(uint256 amount0Out, uint256 amount1Out, address to, bytes calldata data);
+    }
+}
```
Comment (0xKitsune) [inline src/amm/balancer_v2/mod.rs:33]: Does this need to be here?

**Review by 0xKitsune — APPROVED** [review]
Nice work on this! Lets also add `BalancerV2Pool` here as well.

`amm!(UniswapV2Pool, UniswapV3Pool, ERC4626Vault, BalancerV2Pool);`

File: src/amm/balancer_v2/mod.rs [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+    /// Calculates a f64 representation of base token price in the AMM.
+    /// **********************************************************************************************
+    /// calcSpotPrice                                                                             //
+    /// sP = spotPrice                                                                            //
+    /// bI = tokenBalanceIn                ( bI / wI )         1                                  //
+    /// bO = tokenBalanceOut         sP =  -----------  *  ----------                             //
+    /// wI = tokenWeightIn                 ( bO / wO )     ( 1 - sF )                             //
+    /// wO = tokenWeightOut                                                                       //
+    /// sF = swapFee                                                                              //
+    ///**********************************************************************************************/
```
Comment (0xKitsune) [inline src/amm/balancer_v2/mod.rs:92]: Nice touch with these docs!

File: src/amm/balancer_v2/mod.rs [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+        &self,
+        base_token: Address,
+        quote_token: Address,
+    ) -> Result<f64, ArithmeticError> {
+        // Grab the indices of the tokens
+        let base_token_index = self
+            .tokens
+            .iter()
+            .position(|&r| r == base_token)
+            .expect("Base token not found");
```
Comment (0xKitsune) [inline src/amm/balancer_v2/mod.rs:103]: Lets implement error handling for this (and all similar instances).

File: src/amm/balancer_v2/factory.rs [hunk 4]
```diff
@@ -0,0 +1,176 @@
...
+        _block_number: Option<u64>,
+        provider: Arc<P>,
+    ) -> Result<(), AMMError>
+    where
+        T: Transport + Clone,
+        N: Network,
+        P: Provider<T, N>,
+    {
+        // Max batch size for call
+        let step = 127;
```
[the remark this answers is not in the document]
Comment (0xOsiris) [reply src/amm/balancer_v2/factory.rs:83]: I have not - there's quite a bit that needs to be tested including this. Just based off v2 seemed like a similar ish amount of requests happening

[the remark this answers is not in the document]
Comment (0xKitsune) [reply src/amm/balancer_v2/factory.rs:79]: Sounds good, lets just add a todo for a future PR.

**Pull request merged**